File: src/main/js/funk/option/Some.js
funk.option = funk.option || {};
funk.option.Some = (function(){
    "use strict";
    var SomeImpl = function(value){
        funk.option.Option.call(this);

        this._value = value;
    };
    SomeImpl.prototype = new funk.option.Option();
    SomeImpl.prototype.constructor = SomeImpl;
    SomeImpl.prototype.getOrElse = function(f) {
        return this.get();
    };
    SomeImpl.prototype.orElse = function(f){
        return this;
    };
    SomeImpl.prototype.equals = function(that) {
        if(that instanceof funk.option.Option) {
            if(that.isDefined) {
                return funk.util.eq(this.get(), that.get());
            }
        }
        return false;
    };
    SomeImpl.prototype.filter = function(f){
        return f(this.get()) === true ? this : none;
    };
    SomeImpl.prototype.foreach = function(f){
        f(this.get());
    };
    SomeImpl.prototype.flatMap = function(f){
        return funk.util.verifiedType(f(this.get()), funk.option.Option);
    };
    SomeImpl.prototype.map = function(f){
        return funk.option.some(f(this.get()));
    };
    SomeImpl.prototype.get = function(){
        return this._value;
    };
    SomeImpl.prototype.productPrefix = function(){
        return "Some";
    };
    SomeImpl.prototype.productArity = function(){
        return 1;
    };
    SomeImpl.prototype.productElement = function(index){
        if(index == 0) {
            return this.get();
        }
        throw new funk.error.RangeError();
    };
    SomeImpl.prototype.name = "Some";
    SomeImpl.prototype.isEmpty = false;
    SomeImpl.prototype.isDefined = true;
    return SomeImpl;
})();

funk.option.some = function(value){
    return new funk.option.Some(value);
};

// Alias
var some = funk.option.some;

File: src/main/dart/funk/option/some.dart
class Some<T> extends Product<T> implements Option<T> {
  
  T _value;
  
  Some(final T value){
    _value = value;
  }
    
  T getOrElse(final Function f) => _value;

  int hashCode() {
    T h = get;
    if(h is Hashable) {
      Hashable hash = h;
      return hash.hashCode();
    } 
    return 0;
  }
  
  bool equals(final IFunkObject that) {
    if (that is Option) {
      Option<T> option = that;
      if(option.isDefined) {
        return eq(get, option.get);
      }
    }

    return false;
  }

  void foreach(final Function func) {
    func(get);
  }

  Option<T> filter(final Function func) => func(get) == true ? this : none();
  
  Option<T> flatMap(final Function func) => func(get);

  Option<T> map(final Function func) => some(func(get));

  Option<T> orElse(final Function func) => this;
  
  T get get() => _value;
  
  bool get isDefined() => true;

  bool get isEmpty() => false;
  
  int get productArity() => 1;

  String get productPrefix() {
    return "Some";
  }
  
  T productElement(final int index){
    if(index == 0) {
      return get;
    }
    throw new RangeError.empty();
  }
}


some(final value) => new Some(value);